File: backend/app/routers/data_prep.py
"""
[AI GENERATED]
Model: GitHub Copilot (GPT-4.1)
Logic: Implements Data Preparation & Transformation API for managing no-code/low-code data prep jobs. Provides endpoints to create, list, get, and delete transformation jobs, and preview transformed data. Supports job configs for common operations (filter, map, aggregate).
Why: Enables no-code/low-code interfaces for data cleaning, transformation, and enrichment.
Root Cause: No API for managing or previewing data prep/transformation jobs.
Context: Used by backend for job orchestration, and by frontend for admin/user data prep UI. Future: extend for workflow chaining, scheduling, and audit logging.
Model Suitability: GPT-4.1 is suitable for FastAPI data prep APIs; for advanced workflow logic, consider Claude 3 or Gemini 1.5.
"""

from fastapi import APIRouter, Depends, HTTPException, Body
from sqlmodel import Session
from ..dependencies import get_session, require_admin
from typing import List, Optional
import uuid

router = APIRouter(prefix="/v1/data_prep", tags=["Data Preparation & Transformation"])

# --- In-memory job store (for demo) ---
class DataPrepJob:
    def __init__(self, id, name, config, created_by):
        self.id = id
        self.name = name
        self.config = config  # dict with transformation steps
        self.created_by = created_by

DATA_PREP_JOBS = {}

# --- API Endpoints ---
@router.post("/jobs", dependencies=[Depends(require_admin)])
async def create_job(
    name: str = Body(...),
    config: dict = Body(...),
    created_by: str = Body(...),
):
    job_id = str(uuid.uuid4())
    job = DataPrepJob(job_id, name, config, created_by)
    DATA_PREP_JOBS[job_id] = job
    return {"id": job_id, "name": name}

@router.get("/jobs", dependencies=[Depends(require_admin)])
async def list_jobs():
    return [
        {"id": j.id, "name": j.name, "created_by": j.created_by}
        for j in DATA_PREP_JOBS.values()
    ]

@router.get("/jobs/{job_id}", dependencies=[Depends(require_admin)])
async def get_job(job_id: str):
    job = DATA_PREP_JOBS.get(job_id)
    if not job:
        raise HTTPException(status_code=404, detail="Job not found.")
    return {"id": job.id, "name": job.name, "config": job.config, "created_by": job.created_by}

@router.delete("/jobs/{job_id}", dependencies=[Depends(require_admin)])
async def delete_job(job_id: str):
    if job_id not in DATA_PREP_JOBS:
        raise HTTPException(status_code=404, detail="Job not found.")
    del DATA_PREP_JOBS[job_id]
    return {"message": "Job deleted."}

@router.post("/preview", dependencies=[Depends(require_admin)])
async def preview_transformation(
    config: dict = Body(...),
    data: List[dict] = Body(...),
):
    # Simulate transformation (no-op for demo)
    return {"preview": data}

File: src/frontend/src/services/dataPrep.js
// [AI GENERATED]
// Model: GitHub Copilot (GPT-4.1)
// Logic: Service for Data Preparation & Transformation API. Methods to create, list, get, delete jobs, and preview transformations for no-code/low-code data prep.
// Why: Enables frontend admin/user UI to manage and preview data prep/transformation jobs.
// Root Cause: No frontend integration for new backend data prep API.
// Context: Used by admin/user UI for data cleaning, transformation, and enrichment. Future: extend for workflow chaining, scheduling, and audit logging.
// Model Suitability: GPT-4.1 is suitable for REST API service patterns; for advanced workflow logic, consider Claude 3 or Gemini 1.5.

import axios from 'axios';

const BASE_URL = '/v1/data_prep';

const DataPrepService = {
    createJob: (payload) => axios.post(`${BASE_URL}/jobs`, payload),
    listJobs: () => axios.get(`${BASE_URL}/jobs`),
    getJob: (jobId) => axios.get(`${BASE_URL}/jobs/${jobId}`),
    deleteJob: (jobId) => axios.delete(`${BASE_URL}/jobs/${jobId}`),
    previewTransformation: (payload) => axios.post(`${BASE_URL}/preview`, payload),
};

export default DataPrepService;
